### PR TITLE
make res of trace_rho_op real if op + rho are hermitian

### PR DESCRIPTION
**Pull request by the-hampel**
* add function to check if a triqs many body operator is hermitian
* check if input density matrix and operator to trace_rho_op are hermitian and return real part of averaged value

### CODE REVIEW

**Comment by the-hampel** [comment]
We could expose the function `is_op_hermitian` to python?

File: c++/triqs/atom_diag/impl/functions.cpp [hunk 1]
```diff
@@ -76,7 +76,14 @@ namespace triqs {
           if (b_m.first == sp) result += x.coef * trace(b_m.second * density_matrix[sp]);
         }
       }
-      return result;
+      // check if op && density matrix is hermitian then return real(result)
```
Comment (Wentzell) [inline c++/triqs/atom_diag/impl/functions.cpp]:
```suggestion
      // check if op && density matrix are hermitian, if so return real(result)
```

File: c++/triqs/atom_diag/impl/functions.cpp [hunk 2]
```diff
@@ -76,7 +76,14 @@ namespace triqs {
...
+      auto is_block_matrix_hermitian = [](ATOM_DIAG_T::block_matrix_t const &bm) {
+        return std::all_of(bm.begin(), bm.end(), [](ATOM_DIAG_T::matrix_t const &mat) { return max_element(abs(mat - dagger(mat))) == 0.0; });
+      };
+      if (operators::is_op_hermitian(op) && is_block_matrix_hermitian(density_matrix)) {
```
Comment (Wentzell) [inline c++/triqs/atom_diag/impl/functions.cpp]: Argument dependent lookup should trigger here
```suggestion
      if (is_op_hermitian(op) && is_block_matrix_hermitian(density_matrix)) {
```

**Review by Wentzell — CHANGES_REQUESTED** [review]

File: c++/triqs/atom_diag/impl/functions.cpp [hunk 3]
```diff
@@ -76,7 +76,14 @@ namespace triqs {
           if (b_m.first == sp) result += x.coef * trace(b_m.second * density_matrix[sp]);
         }
       }
-      return result;
+      // check if op && density matrix are hermitian, if so return real(result)
+      auto is_block_matrix_hermitian = [](ATOM_DIAG_T::block_matrix_t const &bm) {
+        return std::all_of(bm.begin(), bm.end(), [](ATOM_DIAG_T::matrix_t const &mat) { return max_element(abs(mat - dagger(mat))) == 0.0; });
+      };
+      if (is_op_hermitian(op) && is_block_matrix_hermitian(density_matrix)) {
+        return std::real(result);
```
Comment (Wentzell) [inline c++/triqs/atom_diag/impl/functions.cpp:91]: Maybe you can add a small comment as to why this is even necessary?

Comment (the-hampel) [reply c++/triqs/atom_diag/impl/functions.cpp:91]: this commit below

**Review by Wentzell — APPROVED** [review]
Thank you @the-hampel for this improvement!

**Pull request merged**